### PR TITLE
Vertex declaration changes

### PR DESCRIPTION
**Pull request by larnin**
Allow the user to set his vertex declaration with Primitives, and check if the type of a component on VertexDeclaration and ParticleDeclaration is the same as requested.

### CODE REVIEW

File: include/Nazara/Utility/VertexDeclaration.inl [hunk 1]
```diff
@@ -15,6 +15,17 @@ namespace Nz
 
 		return object.release();
 	}
+
+	template<typename T> 
+	bool VertexDeclaration::HaveComponent(VertexComponent component) const
```
Comment (DrLynix) [inline include/Nazara/Utility/VertexDeclaration.inl]: HasComponent*

Comment (DrLynix) [reply include/Nazara/Utility/VertexDeclaration.inl]: Plus, I think we should have two methods here, "HasComponent" and "HasComponentOfType<T>".

File: src/Nazara/Graphics/ParticleDeclaration.cpp [hunk 2]
```diff
@@ -287,7 +287,7 @@ namespace Nz
 			// ParticleLayout_Billboard : ParticleStruct_Billboard
 			declaration = &s_declarations[ParticleLayout_Billboard];
 			declaration->EnableComponent(ParticleComponent_Color,    ComponentType_Color,  NazaraOffsetOf(ParticleStruct_Billboard, color));
-			declaration->EnableComponent(ParticleComponent_Life,     ComponentType_Int1,   NazaraOffsetOf(ParticleStruct_Billboard, life));
+			declaration->EnableComponent(ParticleComponent_Life,     ComponentType_Float1,   NazaraOffsetOf(ParticleStruct_Billboard, life));
```
Comment (DrLynix) [inline src/Nazara/Graphics/ParticleDeclaration.cpp]: image

File: include/Nazara/Utility/Algorithm.hpp [hunk 3]
```diff
@@ -57,6 +60,49 @@ namespace Nz
 	NAZARA_UTILITY_API void SkinPositionNormalTangent(const SkinningData& data, unsigned int startVertex, unsigned int vertexCount);
 
 	NAZARA_UTILITY_API void TransformVertices(VertexPointers vertexPointers, unsigned int vertexCount, const Matrix4f& matrix);
+	
+
+	template <typename T, std::enable_if_t<std::is_same<std::remove_const_t<T>, Color>::value>...>
+	ComponentType ComponentTypeOf() { return ComponentType_Color; }
```
Comment (DrLynix) [inline include/Nazara/Utility/Algorithm.hpp]: I would prefer it named "GetComponentTypeOf"

Comment (DrLynix) [reply include/Nazara/Utility/Algorithm.hpp]: Also, it should have a default implementation with a static_assert(false) saying something like "missing implementation for type"

File: src/Nazara/Graphics/ParticleDeclaration.cpp [hunk 4]
```diff
@@ -298,7 +298,7 @@ namespace Nz
 
 			// ParticleLayout_Model : ParticleStruct_Model
 			declaration = &s_declarations[ParticleLayout_Model];
-			declaration->EnableComponent(ParticleComponent_Life,     ComponentType_Int1,       NazaraOffsetOf(ParticleStruct_Model, life));
+			declaration->EnableComponent(ParticleComponent_Life,     ComponentType_Float1,       NazaraOffsetOf(ParticleStruct_Model, life));
```
Comment (DrLynix) [inline src/Nazara/Graphics/ParticleDeclaration.cpp]: Same here

File: src/Nazara/Graphics/ParticleDeclaration.cpp [hunk 5]
```diff
@@ -308,7 +308,7 @@ namespace Nz
 			// ParticleLayout_Sprite : ParticleStruct_Sprite
 			declaration = &s_declarations[ParticleLayout_Sprite];
 			declaration->EnableComponent(ParticleComponent_Color,    ComponentType_Color,  NazaraOffsetOf(ParticleStruct_Sprite, color));
-			declaration->EnableComponent(ParticleComponent_Life,     ComponentType_Int1,   NazaraOffsetOf(ParticleStruct_Sprite, life));
+			declaration->EnableComponent(ParticleComponent_Life,     ComponentType_Float1,   NazaraOffsetOf(ParticleStruct_Sprite, life));
```
Comment (DrLynix) [inline src/Nazara/Graphics/ParticleDeclaration.cpp]: Same here

**Comment by Gawaboumga** [comment]
The compilation error seems difficult to get rid off at first sight. I will try to see what can be achieved through "include-what-you-use" tool.

File: include/Nazara/Utility/Algorithm.hpp [hunk 6]
```diff
@@ -8,17 +8,26 @@
...
 namespace Nz
 {
+	class Joint;
+
+	class VertexStruct_XYZ_Normal_UV_Tangent_Skinning;
+	typedef VertexStruct_XYZ_Normal_UV_Tangent_Skinning SkeletalMeshVertex;
+
+	class VertexStruct_XYZ_Normal_UV_Tangent;
+	typedef VertexStruct_XYZ_Normal_UV_Tangent MeshVertex;
+
```
Comment (DrLynix) [inline include/Nazara/Utility/Algorithm.hpp]: Could you please put all the forward declaration together and put all of typedef (btw, using would be better) together too?

File: include/Nazara/Utility/Algorithm.hpp [hunk 7]
```diff
@@ -57,6 +66,36 @@ namespace Nz
...
+	template<> constexpr ComponentType componentTypeId<Vector3f>() { return ComponentType_Float3; }
+	template<> constexpr ComponentType componentTypeId<Vector4f>() { return ComponentType_Float4; }
+	template<> constexpr ComponentType componentTypeId<int>() { return ComponentType_Int1; }
+	template<> constexpr ComponentType componentTypeId<Vector2i>() { return ComponentType_Int2; }
+	template<> constexpr ComponentType componentTypeId<Vector3i>() { return ComponentType_Int3; }
+	template<> constexpr ComponentType componentTypeId<Vector4i>() { return ComponentType_Int4; }
+	template<> constexpr ComponentType componentTypeId<Quaternionf>() { return ComponentType_Quaternion; }
+
+	template<typename T>
+	constexpr const ComponentType GetComponentTypeOf() { return componentTypeId<std::decay_t<T>>(); }
```
Comment (DrLynix) [inline include/Nazara/Utility/Algorithm.hpp]: No code in the .hpp, please put all the template specializations and function code inside the .inl file

File: include/Nazara/Utility/Algorithm.hpp [hunk 8]
```diff
@@ -57,6 +66,36 @@ namespace Nz
...
 
 	NAZARA_UTILITY_API void TransformVertices(VertexPointers vertexPointers, unsigned int vertexCount, const Matrix4f& matrix);
+	
+	template<typename T>
+	struct IsSuitableForComponent
+	{
+		constexpr static bool value = false;
+	};
+
+	template<class T> constexpr ComponentType componentTypeId()
```
Comment (DrLynix) [inline include/Nazara/Utility/Algorithm.hpp]: No camelCase for functions

File: include/Nazara/Utility/Algorithm.hpp [hunk 9]
```diff
@@ -66,36 +65,16 @@ namespace Nz
 	NAZARA_UTILITY_API void SkinPositionNormalTangent(const SkinningData& data, unsigned int startVertex, unsigned int vertexCount);
 
 	NAZARA_UTILITY_API void TransformVertices(VertexPointers vertexPointers, unsigned int vertexCount, const Matrix4f& matrix);
-	
+
 	template<typename T>
 	struct IsSuitableForComponent
```
Comment (DrLynix) [inline include/Nazara/Utility/Algorithm.hpp]: This should be moved to the .inl too (as it has no use for users) in a Detail namespace

**Pull request merged**